File: fsharp/space-age/SpaceAge.fs
module SpaceAge

type SpaceAge(seconds : decimal) =

  let secondsPerYear =
    [
      "earth", 31557600m;
      "mercury", 7600543.81992m;
      "venus", 19414149.052176m;
      "mars", 59354032.69008m;
      "jupiter", 374355659.124m;
      "saturn", 929292362.8848m;
      "uranus", 2651370019.3296m;
      "neptune", 5200418560.032m;
    ] |> Map.ofList

  let round2dp x = System.Math.Round(x |> float, 2)

  let planetaryYears planet =
    round2dp <| (seconds / secondsPerYear.[planet]);

  member this.Seconds = seconds

  member this.onEarth = planetaryYears "earth"

  member this.onMercury = planetaryYears "mercury"

  member this.onVenus = planetaryYears "venus"

  member this.onMars = planetaryYears "mars"

  member this.onSaturn = planetaryYears "saturn"

  member this.onJupiter = planetaryYears "jupiter"

  member this.onNeptune = planetaryYears "neptune"

  member this.onUranus = planetaryYears "uranus"

File: ruby/space-age/space_age.rb
class SpaceAge
  attr_reader :seconds

  def initialize(seconds)
    @seconds = seconds
  end

  private

  def method_missing(symbol)
    name = symbol.to_s
    super unless name.start_with?('on_')
    planet = name[3..-1].to_sym
    year_length(planet) || super
  end

  def year_length(planet)
    return unless year_lengths.key?(planet)
    seconds / year_lengths[planet]
  end

  def year_lengths
    {
      mercury: 7_600_543.81992,
      venus: 19_414_149.052176,
      earth: 31_557_600.0,
      mars: 59_354_032.69008,
      jupiter: 374_355_659.124,
      saturn: 929_292_362.8848,
      uranus: 2_651_370_019.3296,
      neptune: 5_200_418_560.032
    }
  end
end
